Add support for AIX

mhdawson opened the pull request:
For AIX we need to use gmake

For AIX we need to set up the path to the exp file
which contains the symbols needed for linking.
The file will either be in one of the following
depeding on whether are are in installed or
development environment:
  - the include/node directory
  - the out/Release directory

The work to add the required node.exp file is being covered in this PR https://github.com/nodejs/node/pull/3114

mhdawson commented:
I've tested this in a build of Node from master.  I've looked through the readme for node-gyp to see how changes are normally tested but I've not found any instructions. Can somebody point me at were it is documented what testing should be done ?

rvagg commented on lib/build.js:
```diff
@@ -21,6 +21,7 @@ exports.usage = 'Invokes `' + (win ? 'msbuild' : 'make') + '` and builds the mod
 function build (gyp, argv, callback) {
   var release = processRelease(argv, gyp, process.version, process.release)
     , makeCommand = gyp.opts.make || process.env.MAKE
+      || (process.platform.indexOf('aix') != -1 && 'gmake')
```
the mixed use of `&&` and the ternary operator below makes this a bit confusing, the duplicate resolving of `gmake` suggests redundancy, and, shouldn't it just be `aix` when it's AIX? the `indexOf` is for the *bsd variants I _think_.

Can I suggest squashing both these lines to: `((/(^aix$)|bsd/).test(process.platform) ? 'gmake' : 'make')`

bnoordhuis replied:
Maybe platform-specific overrides are better done in parseArgv / parseOpts in lib/node-gyp.js.

mhdawson replied:
it is kind of of ugly.  What do you think of this. It's more verbose but also more obvious whats going on 

function build (gyp, argv, callback) {
  var platformMake = 'make';
  if (process.platform == 'aix') {
    platformMake = 'gmake';
  } else if (process.platform.indexOf('bsd') != -1) {
    platformMake = 'gmake');
  }

  var release = processRelease(argv, gyp, process.version, process.release)
    , makeCommand = gyp.opts.make || process.env.MAKE || platformMake
    , command = win ? 'msbuild' : makeCommand
    , buildDir = path.resolve('build')
    , configPath = path.resolve(buildDir, 'config.gypi')
    , jobs = gyp.opts.jobs || process.env.JOBS
    , buildType
    , config
    , arch
    , nodeDir
    , copyDevLib

rvagg commented on lib/configure.js:
```diff
@@ -161,6 +161,42 @@ function configure (gyp, argv, callback) {
 
+  function findNodeRootDirectory() {
+    // Have a look to see what is above us, to try and work out where we are
+    npm_parent_directory = path.join(__dirname, '..', '..','..','..')
```
`path.join(__dirname, '../../../..')` should be fine here

mhdawson replied:
ok

rvagg commented on lib/configure.js:
```diff
@@ -161,6 +161,42 @@ function configure (gyp, argv, callback) {
+    log.verbose('node-gyp root', 'npm_parent_directory is ' + path.basename(npm_parent_directory))
+    node_root_dir = ""
+
+    log.verbose('node-gyp root','Finding node root directory')
```
space after `,`

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -161,6 +161,42 @@ function configure (gyp, argv, callback) {
+    log.verbose('node-gyp root','Finding node root directory')
+    if (path.basename(npm_parent_directory) == 'lib') {
+      // We are in a node install directory
+      node_root_dir = path.join(npm_parent_directory, '..', '..')
```
`'../..'`

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -161,6 +161,42 @@ function configure (gyp, argv, callback) {
+      log.verbose('node-gyp root', 'in install directory, root = ' + node_root_dir)
+    } else if (path.basename(npm_parent_directory) == 'node_modules') {
+      // We are in a node build directory, where npm lives under lib/
+      node_root_dir = path.join(npm_parent_directory, '..', '..', 'include', 'node')
```
`path.join(npm_parent_directory, '../../include/node')`

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -161,6 +161,42 @@ function configure (gyp, argv, callback) {
+        if (process.platform == 'win32') {
+          node_root_dir = path.join(node_dir, '..')
+        } else {
+          node_root_dir = path.join(node_dir, '..', '..')
```
'../..`

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -319,6 +374,9 @@ function configure (gyp, argv, callback) {
       argv.push('-Dvisibility=default')
       argv.push('-Dnode_root_dir=' + nodeDir)
+      if (process.platform == 'aix') {
+        argv.push('-Dnode_exp_file=' + node_exp_file )
```
remove space before `)`

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -299,6 +335,25 @@ function configure (gyp, argv, callback) {
+    // development environment:
+    //  - the include/node directory
+    //  - the out/Release directory
+    var node_exp_file = "";
```
single quotes, no semicolon `= ''`

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -299,6 +335,25 @@ function configure (gyp, argv, callback) {
+    //  - the out/Release directory
+    var node_exp_file = "";
+    if (process.platform == 'aix') {
+      var node_root_dir = findNodeRootDirectory();
```
no semicolon

mhdawson replied:
k

rvagg commented on lib/configure.js:
```diff
@@ -299,6 +335,25 @@ function configure (gyp, argv, callback) {
+    var node_exp_file = "";
+    if (process.platform == 'aix') {
+      var node_root_dir = findNodeRootDirectory();
+      node_exp_file = path.resolve(node_root_dir, 'include/node/node.exp')
```
can you confirm these are included in the header tarballs created upstream with your recent changes?

rvagg commented:
I'm pretty sure the `findNodeRootDirectory()` approach is wrong here and makes too many unsafe assumptions. Can you have a look at how it finds `config.gypi` and emulate that instead? I understand the problem of finding node.exp when using `--nodedir` but that's only a minor case, the most important one being finding the file in `~/.node-gyp/version/include/node/node.exp`.

rvagg commented:
re tests, there isn't much in the way of tests here. There are some docker tests you can run in test/docker.sh but obviously only for Linux but it's worth making sure you haven't broken anything else. Other testing has to be manual, which can be quite tricky when we haven't got any headers tarballs on nodejs.org that you can trial with! Once we get an AIX machine (or two, or three, whatever) in CI then we could produce tarballs for releases || nightlies with node.exp in it and test with that.

But, that raises another question, is node.exp specific to the arch used? Does it need a new one for BE, LE, 32-bit, 64-bit? If so, then it can't go into the headers tarball and you're going to need to scrap a lot of this code and emulate the node.lib download process we have for Windows and we're going to need to set up new directories in nodejs.org/download/ to contain node.exp for all of the different architectures.

rvagg commented:
OK, another rethink on this, backing up on what I've just said I realise your approach now is to find `node.exp` as it's been installed on the system along with `node` from a binary tarball or some other official package, not necessarily have it in the headers or download from nodejs.org like `node.lib`. This is something that many of us have talked about for some time now wrt general headers because they are installed on many systems but node-gyp will always download them again anyway. A disparity that's existed since 0.10.

In that case, perhaps this is the right approach, _but_ I personally don't want to see this logic baked in to node-gyp and would rather see it split off to a separate module that can locate the Node `include` dir on any system if it exists. That way it can be independently tested and iterated on rather than becoming part of the big logic puzzle that node-gyp currently is. I know that's not the expedient path here but as it is there's no clear way to test this logic in here. Perhaps a middle ground is to do what I did with `process-release.js` (which I would also like to extract to a separate package) and treat it as independent and test it independently.

bnoordhuis commented on lib/configure.js:
```diff
@@ -299,6 +335,25 @@ function configure (gyp, argv, callback) {
+    if (process.platform == 'aix') {
+      var node_root_dir = findNodeRootDirectory();
+      node_exp_file = path.resolve(node_root_dir, 'include/node/node.exp')
+      if (!fs.existsSync(node_exp_file)) {
```
`fs.existsSync()` emits deprecation warnings.

mhdawson replied:
Noted will replace with one of the suggested alternates

bnoordhuis commented on lib/configure.js:
```diff
@@ -299,6 +335,25 @@ function configure (gyp, argv, callback) {
+      var node_root_dir = findNodeRootDirectory();
+      node_exp_file = path.resolve(node_root_dir, 'include/node/node.exp')
+      if (!fs.existsSync(node_exp_file)) {
+        node_exp_file = path.resolve(node_root_dir, 'out/Release/node.exp')
```
What about debug builds?

mhdawson replied:
Good point, will add line that looks in out/Debug  as well

mhdawson commented:
Sorry for taking so long to get back to this one.  First part in Node is now in so will be moving on to this one soon.  I am out at a conference this week but it will my priority after that.

mhdawson commented:
Have updates to reflect the comments in my local repo, just need to test now

mhdawson commented:
@bnoordhuis  @rvagg  updated to address review comments

bnoordhuis commented on lib/build.js:
```diff
@@ -19,9 +19,15 @@ var fs = require('graceful-fs')
+  if (process.platform == 'aix') {
+    platformMake = 'gmake';
+  } else if (process.platform.indexOf('bsd') != -1) {
+    platformMake = 'gmake';
```
Style nits: no trailing semicolons (for better or worse)

mhdawson replied:
k

bnoordhuis commented on lib/configure.js:
```diff
@@ -299,6 +300,34 @@ function configure (gyp, argv, callback) {
+                        'out/Release/node.exp',
+                        'out/Debug/node.exp',
+                        'node.exp']
+      for(var next in candidates) {
```
Style nit: space after `for`, ditto for the `catch` below.

mhdawson replied:
k

bnoordhuis commented on lib/find-node-directory.js:
```diff
@@ -0,0 +1,40 @@
+    var node_dir = path.dirname(process.execPath)
+    var directory_up = path.basename(node_dir)
+    if (directory_up == 'bin') {
+      node_root_dir  = path.join(node_dir, '..')
```
Style nit: one space too many

bnoordhuis commented on lib/find-node-directory.js:
```diff
@@ -0,0 +1,40 @@
+
+  log.verbose('node-gyp root', 'Finding node root directory')
+  if (path.basename(npm_parent_directory) == 'lib') {
+    // We are in a node install directory
```
I'm not sure I understand when this would happen.   With a path like `/opt/node/lib/node_modules/npm/node_modules/node-gyp/lib`, you arrive at `/opt/node/lib/node_modules`, not `/opt/node/lib`.

mhdawson replied:
From the history I can gather the function to get the Node directory used to exist somewhere else as well but no longer does.  Looking at the potential installation configurations I agree that 'lib' does not make sense.  The cases I've been testing include when you build from source and have that layout and the case were you install from the tar.gz.  From those I think we should be checking 2 cases. One where the parent is "node_modules" (which is covered) and one where the parent is "deps". I'll update and add more comments to try to make it clear which cases are being covered.

bnoordhuis commented on lib/find-node-directory.js:
```diff
@@ -0,0 +1,40 @@
+    // Else return the default blank, "".
+  }
+  return node_root_dir
+}
```
Can you add a few regression tests for this function?  I'd make it so that you can plug in a mock `__dirname` and `process` object.

bnoordhuis commented on lib/find-node-directory.js:
```diff
@@ -0,0 +1,40 @@
+function findNodeDirectory() {
+  // Have a look to see what is above us, to try and work out where we are
+  npm_parent_directory = path.join(__dirname, '../../../..')
+  log.verbose('node-gyp root', 'npm_parent_directory is ' + path.basename(npm_parent_directory))
```
Style nit: can you try to keep lines <= 80 columns?  I know node-gyp isn't very consistent in that respect but I'd like new code to not have overly long lines.

mhdawson replied:
my bad, need to be more diligent about that

mhdawson commented:
@bnoordhuis updated to address your latest comments. Added the unit tests and took a closer look as you had found a place were the code was obviously stale.  I fixed up the initial checks to be for 'node_modules' and 'deps' and validated directories on windows and fixed a few issues there as well.

bnoordhuis commented on test/test-find-node-directory.js:
```diff
@@ -0,0 +1,98 @@
+    t.equal(
+      findNodeDirectory('/y/lib/node_modules/npm/node_modules/node-gyp/lib',
+                         processObj), path.join('/y'));
+  }
```
Can you test both branches here?  I.e., rewrite it so it's called once with `processObj.platform === 'win32'` and once without?

bnoordhuis replied:
I'd recommend doing that everywhere where you call findNodeDirectory() with `{ platform: process.platform }`.  The idea is that all code paths are tested on all platforms.

bnoordhuis commented on lib/build.js:
```diff
@@ -19,9 +19,15 @@ var fs = require('graceful-fs')
+  var platformMake = 'make'
+  if (process.platform === 'aix') {
+    platformMake = 'gmake'
+  } else if (process.platform.indexOf('bsd') != -1) {
```
Can you change this to `!==` while you're here?

mhdawson replied:
sure

bnoordhuis commented on lib/configure.js:
```diff
@@ -299,6 +300,34 @@ function configure (gyp, argv, callback) {
+           fs.accessSync(node_exp_file, fs.R_OK)
+           // exp file found, stop looking
+           break
+         } catch(exception) {
```
Micro-nit: space after `catch`.

mhdawson replied:
k

bnoordhuis commented on test/test-find-node-directory.js:
```diff
@@ -0,0 +1,98 @@
+  var processObj = {execPath: '/x/y/bin/node', platform: process.platform}
+  t.equal(
+    findNodeDirectory('/x/deps/npm/node_modules/node-gyp/lib', processObj),
+                      path.join('/x'));
```
There are a few stray semicolons in this file.

mhdawson replied:
Thanks its hard to stomp out that habit.

thefourtheye commented on lib/find-node-directory.js:
```diff
@@ -0,0 +1,61 @@
+      node_root_dir = path.join(npm_parent_directory, '../..')
+    }
+    log.verbose('node-gyp root', 'in install directory, root = '
+    + node_root_dir)
```
Minor nit: Shouldn't the `+` be on the previous line?

mhdawson replied:
Is there a standard for this ?   I should at least line up with the start of the variables like this:

log.verbose('node-gyp root', 'in install directory, root = '
                  + node_root_dir)

like I have done elsewhere.  I'll make that change

mhdawson commented:
@bnoordhuis @thefourtheye updated to address latest comments

bnoordhuis commented:
LGTM.  @thefourtheye WDYT?

thefourtheye commented on lib/configure.js:
```diff
@@ -299,6 +300,34 @@ function configure (gyp, argv, callback) {
+    // for AIX we need to set up the path to the exp file
+    // which contains the symbols needed for linking.
+    // The file will either be in one of the following
+    // depeding on whether are are in installed or
```
depending on whether it is

mhdawson replied:
fixed

thefourtheye commented on lib/configure.js:
```diff
@@ -299,6 +300,34 @@ function configure (gyp, argv, callback) {
+                        'out/Release/node.exp',
+                        'out/Debug/node.exp',
+                        'node.exp']
+      for (var next in candidates) {
```
Iterating an array with `for...in` is generally not advisable.

bnoordhuis replied:
Doesn't really matter for a test though.

thefourtheye replied:
But this is not a test :O

mhdawson replied:
Changed here and in the test files

thefourtheye commented on lib/configure.js:
```diff
@@ -319,6 +348,9 @@ function configure (gyp, argv, callback) {
       argv.push('-Dvisibility=default')
       argv.push('-Dnode_root_dir=' + nodeDir)
+      if (process.platform === 'aix') {
+        argv.push('-Dnode_exp_file=' + node_exp_file)
```
If the file is not found, should we still add this flag?

mhdawson replied:
I'd prefer that we still add the flag so that we can see what occurred.  In verbose mode we'll be able to see what the value for -Dnode_exp_file is

thefourtheye replied:
Oh okay then

thefourtheye commented on test/test-find-node-directory.js:
```diff
@@ -0,0 +1,115 @@
+    var processObj = {execPath: '/x/y/bin/node', platform: currentPlatform}
+    t.equal(
+      findNodeDirectory('/x/deps/npm/node_modules/node-gyp/lib', processObj),
+                        path.join('/x'))
```
`path.join` is used all over the tests but it actually doesn't join paths. Am I totally missing the point of its usage?

mhdawson replied:
without the join the patch ends up being incorrect on windows, the join fixes up the '/' to be appropriate for the platform

thefourtheye replied:
But you are hardcoding them right? I need to check if the join changes the slashes

thefourtheye commented on test/test-find-node-directory.js:
```diff
@@ -0,0 +1,115 @@
+
+// we should find the directory based on the directory
+// the script is running in and it should match the layout
+// in a installed tree where npm is installed in
```
minor nit: `in an installed `

mhdawson replied:
fixed

thefourtheye commented on test/test-find-node-directory.js:
```diff
@@ -0,0 +1,115 @@
+  }
+})
+
+// we shoudl not find it as it will not match based on the execPath nor
```
should

mhdawson replied:
fixed

thefourtheye commented:
Have few minor comments but overall change LGTM

mhdawson commented:
@thefourtheye addressed your comments.  If it looks ok to you I can squash into one commit

thefourtheye commented on lib/configure.js:
```diff
@@ -299,6 +300,34 @@ function configure (gyp, argv, callback) {
+           fs.accessSync(node_exp_file, fs.R_OK)
+           // exp file found, stop looking
+           break
+         } catch (exception) {
```
If the file is not found anywhere, `node_exp_file` will still be `node.exp` at the end of the loop. Is that okay or you want to reset to empty string here?

mhdawson replied:
I think its ok as if this does occur I don't think its going to be any clearer what the problem is with it set to an empty string.

thefourtheye replied:
Cool, let's land it then. LGTM

mhdawson commented:
Squashed down to a single commit

bnoordhuis commented:
Thanks Michael, landed in 9bfa087.
